### PR TITLE
NVS basic tests

### PR DESCRIPTION
**Pull request by kapi-no**
Added basic tests for the NVS file system. One of the tests covers power down during the flash write operation.

### CODE REVIEW

**Comment by kapi-no** [comment]
CC: @nvlsianpu

**Review by nashif — APPROVED** [review]

File: tests/subsys/fs/nvs/src/main.c [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+	int err = 1;
+	struct flash_pages_info info;
+
+	fs.offset = DT_FLASH_AREA_STORAGE_OFFSET;
+	err = flash_get_page_info_by_offs(device_get_binding(DT_FLASH_DEV_NAME),
+					  fs.offset, &info);
+	zassert_true(err == 0,  "Unable to get page info: %d", err);
+
+	fs.sector_size = info.size;
+	fs.sector_count = 5U;
```
Comment (nvlsianpu) [inline tests/subsys/fs/nvs/src/main.c]: magic value

File: tests/subsys/fs/nvs/src/main.c [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+		memcpy(wr_buf + i, pattern, sizeof(pattern));
+	}
+
+	len = nvs_write(&fs, TEST_DATA_ID, wr_buf, sizeof(wr_buf));
+	zassert_true(len == sizeof(wr_buf), "nvs_write failed: %d", len);
+
+	len = nvs_read(&fs, TEST_DATA_ID, rd_buf, sizeof(rd_buf));
+	zassert_true(len == sizeof(rd_buf),  "nvs_read unexpected failure: %d",
+			len);
+	zassert_true(memcmp(wr_buf, rd_buf, sizeof(rd_buf)) == 0,
```
Comment (nvlsianpu) [inline tests/subsys/fs/nvs/src/main.c]: use zassert_mem_equal() (whole file)

**Review by nvlsianpu — CHANGES_REQUESTED** [review]

File: tests/subsys/fs/nvs/src/main.c [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+	*flash_write_stat = 0;
+
+	/* Flash simulator will lose part of the data at the end of this write.
+	 * This should simulate power down during flash write. The written data
+	 * are corrupted at this point and should be discarded by the NVS.
+	 */
+	len = nvs_write(&fs, TEST_DATA_ID, wr_buf_2, sizeof(wr_buf_2));
+	zassert_true(len == sizeof(wr_buf_2), "nvs_write failed: %d", len);
+
+	len = nvs_read(&fs, TEST_DATA_ID, rd_buf, sizeof(rd_buf));
```
Comment (nvlsianpu) [inline tests/subsys/fs/nvs/src/main.c]: Behavior after subsequent `nvs_init()` should be checked.

**Review by nvlsianpu — CHANGES_REQUESTED** [review]

File: tests/subsys/fs/nvs/src/main.c [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+	zassert_true(len == sizeof(rd_buf),  "nvs_read unexpected failure: %d",
+			len);
+	zassert_true(memcmp(wr_buf_2, rd_buf, sizeof(rd_buf)) != 0,
```
Comment (nvlsianpu) [inline tests/subsys/fs/nvs/src/main.c]: messages should explain what is wrong with the software under test.

File: tests/subsys/fs/nvs/src/main.c [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+	}
+}
+
+void test_nvs_init(void)
+{
+	int err = 1;
+	struct flash_pages_info info;
+
+	fs.offset = DT_FLASH_AREA_STORAGE_OFFSET;
+	err = flash_get_page_info_by_offs(device_get_binding(DT_FLASH_DEV_NAME),
```
[the remark this answers is not in the document]
Comment (kapi-no) [reply tests/subsys/fs/nvs/src/main.c]: okay, but I will need to add flash_map dependency to the test

File: tests/subsys/fs/nvs/src/main.c [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+		stats_reset(sim_thresholds);
+	}
+}
+
+void test_nvs_init(void)
+{
+	int err = 1;
+	struct flash_pages_info info;
+
+	fs.offset = DT_FLASH_AREA_STORAGE_OFFSET;
```
[the remark this answers is not in the document]
Comment (nvlsianpu) [reply tests/subsys/fs/nvs/src/main.c]: after rethinking - its correct here.

**Review by Laczen — APPROVED** [review]
Nice work, I didn't know the `%zx` format specifier, thanks for that.

I would remove the USE_PARTITION_MANAGER defines, I couldn't find pm_config.h in zephyr code.

I would wrap all simulator specific routines in a #ifdef CONFIG_FLASH_SIM (don't know if this exists tough, my knowledge on the flash simulator is limited). This would allow the test also to run on real hardware as well (except of course the test of the write failure).

File: tests/subsys/fs/nvs/src/main.c [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+#include "flash.h"
+#include "flash_map.h"
+#include "stats.h"
+#include "nvs/nvs.h"
+
+#if USE_PARTITION_MANAGER
+#include <pm_config.h>
+#define TEST_FLASH_AREA_STORAGE_OFFSET PM_MCUBOOT_STORAGE_ADDRESS
+#else
+#define TEST_FLASH_AREA_STORAGE_OFFSET DT_FLASH_AREA_STORAGE_OFFSET
```
Comment (nvlsianpu) [inline tests/subsys/fs/nvs/src/main.c]: keep only this line for
```
#define TEST_FLASH_AREA_STORAGE_OFFSET DT_FLASH_AREA_STORAGE_OFFSET
```

rest  are unwanted here.

**Review by nvlsianpu — CHANGES_REQUESTED** [review]

File: tests/subsys/fs/nvs/src/main.c (hunk 7 above)
Comment (kapi-no) [reply tests/subsys/fs/nvs/src/main.c]: fixed

**Comment by kapi-no** [comment]
> I would wrap all simulator specific routines in a #ifdef CONFIG_FLASH_SIM (don't know if this exists tough, my knowledge on the flash simulator is limited). This would allow the test also to run on real hardware as well (except of course the test of the write failure).

Yes, that would be feasible provided that we can elicit corrupted write operation on the real hardware in the test. Here it is achieved because the flash simulator doesn't store any data after it executes a certain number of write calls and this number is configured by the test.

This test scenario is supposed to simulate the power loss when the NVS writes to the flash. It is intended to verify the NVS (not the flash drivers underneath), so IMHO running the test on the real hardware doesn't bring much value since all platforms share the same NVS codebase.

Anyway, thanks for the feedback.

**Comment by Laczen** [comment]
> > I would wrap all simulator specific routines in a #ifdef CONFIG_FLASH_SIM (don't know if this exists tough, my knowledge on the flash simulator is limited). This would allow the test also to run on real hardware as well (except of course the test of the write failure).
> 
> Yes, that would be feasible provided that we can elicit corrupted write operation on the real hardware in the test. Here it is achieved because the flash simulator doesn't store any data after it executes a certain number of write calls and this number is configured by the test.
> 
> This test scenario is supposed to simulate the power loss when the NVS writes to the flash. It is intended to verify the NVS (not the flash drivers underneath), so IMHO running the test on the real hardware doesn't bring much value since all platforms share the same NVS codebase.
> 
> Anyway, thanks for the feedback.

OK, I would rename the PR into: "nvs basic powerfail test"

**Review by nvlsianpu — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**